### ISSUE TITLE
KDTreeSingleIndexDynamicAdaptor::findNeighbors returns out of bound index

### ISSUE THREAD

**Issue by tommasocarella** [jlblancoc/nanoflann#140]
Hi,

I econuntered a problem while performing knn search using nanoflann.
Basically, what I am trying to do is to iteratively perform knn search on a 3D point cloud, adding new points to the kdTree step by step. What happens is that sometimes the algorithm returns a nn index which is out of bound with respect to my point cloud vector, as the images below show:
![image](https://user-images.githubusercontent.com/14278767/106890321-6f8f2f80-66e9-11eb-9fbc-be7e30a876f4.png)
![image](https://user-images.githubusercontent.com/14278767/106890364-833a9600-66e9-11eb-98c6-9c7b4ec0d80f.png)
In fact, using index 40 to access a 0-39 array will cause an "out_of_bound exception"

**Code**
if ((inputPatternsNumber - 1) % (lambda) == 0 && inputPatternsNumber > 1) {
  nanoflannPoint<float> nanoGngNew;
  nanoGngNew.x = (*(nodeVectorPtr.end() - 1))->getNodePosition().x;
  nanoGngNew.y = (*(nodeVectorPtr.end() - 1))->getNodePosition().y;
  nanoGngNew.z = (*(nodeVectorPtr.end() - 1))->getNodePosition().z;
  nanoGngCloud.points.push_back(nanoGngNew);

  index.addPoints(nanoGngCloud.kdtree_get_point_count(), nanoGngCloud.kdtree_get_point_count());
 }

const size_t num_results = 2;
size_t ret_index[num_results];
float out_dist_sqr[num_results];
nanoflann::KNNResultSet<float> resultSet(num_results);
resultSet.init(ret_index, out_dist_sqr);
index.findNeighbors(resultSet, nanoGngPoint, nanoflann::SearchParams(10));

Basically, after some iterations I am adding new points to the point cloud while adding new points to kdTree.

Hope that someone has a solution!
Thank you,

Tommaso

**Comment by jlblancoc** [comment]
I'll try to look at it. Any chance to have a complete CMakeLists.txt + test.cpp so we can reproduce the exact issue? Thanks

**Comment by tommasocarella** [comment]
Hi @jlblancoc,
Thank you so much for your answer.
I think I found the solution to the problem and it was in my code:
`index.addPoints(nanoGngCloud.kdtree_get_point_count(), nanoGngCloud.kdtree_get_point_count());` **became**
`index.addPoints(nanoGngCloud.kdtree_get_point_count()-1, nanoGngCloud.kdtree_get_point_count()-1);`

Basicalli, I was adding a chunk of memory to the index that did not match any point of my point vector.

Thank you again,
Tommaso

**Comment by jlblancoc** [comment]
Fantastic!

Anyway, I'll try to redesign the library to make it as easy to use as possible.... not a priority for current projects, though!
